securityscanutils/securityscan.go

```diff
@@ -0,0 +1,284 @@
+		}
+		if constraint.Check(versionToScan) {
+			// We want to make sure that each version only matches ONE constraint provided
+			// in the constraint -> []images map, so that we are scanning the right images for each version
+			if imagesToScan != nil {
+				return nil, eris.Errorf(
+					"version %s matched more than one constraint provided, please make all constraints mutually exclusive", versionToScan.String())
+			}
+			imagesToScan = images
+		}
+
+	}
+	if imagesToScan == nil {
+		return nil, eris.Errorf("version %s matched no constraints and has no images to scan", versionToScan.String())
+	}
+	return imagesToScan, nil
+}
+
+// Runs trivy scan command
+// returns if trivy scan ran successfully and error if there was one
+func RunTrivyScan(image, version, templateFile, output string) (bool, error) {
+	// Ensure Trivy is installed and on PATH
+	_, err := exec.LookPath("trivy")
+	if err != nil {
+		return false, eris.Wrap(err, "trivy is not on PATH, make sure that the trivy v0.18 is installed and on PATH")
+	}
+	args := []string{"image", "--severity", "HIGH,CRITICAL", "--format", "template", "--template", "@" + templateFile, "--output", output, image}
+	cmd := exec.Command("trivy", args...)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		// delete empty trivy output file that may have been created
+		_ = os.Remove(output)
+		// swallow error if image is not found error, so that we can continue scanning releases
+		// even if some releases failed and we didn't publish images for those releases
+		if IsImageNotFoundErr(string(out)) {
+			log.Warnf("image %s not found for version %s", image, version)
+			return false, nil
+		}
+		return false, eris.Wrapf(err, "error running trivy scan on image %s, version %s, Logs: \n%s", image, version, string(out))
+	}
+	return true, nil
+}
+
+type SarifMetadata struct {
+	Ref       string `json:"ref"`
+	CommitSha string `json:"commit_sha"`
+	Sarif     string `json:"sarif"`
+}
+
+type Response struct {
+	ShaObject `json:"object"`
+}
+
+type ShaObject struct {
+	Sha string `json:"sha"`
+}
+
+func (r *SecurityScanRepo) UploadSecurityScanToGithub(fileName, versionTag string) error {
+	githubRepoApiUrl := fmt.Sprintf("https://api.github.com/repos/%s/%s", r.Owner, r.Repo)
+	githubToken, err := osutils.GetEnvE("GITHUB_TOKEN")
+	if err != nil {
+		return err
+	}
+	resp, err := req.Get(githubRepoApiUrl+"/git/refs/tags/v"+versionTag, req.Header{"Authorization": "token " + githubToken})
+	if err != nil {
+		return eris.Wrapf(err, "Unable to get commit for version v%s", versionTag)
+	}
+	shaResp := &Response{}
+	resp.ToJSON(shaResp)
+	fmt.Printf("%+v\n", shaResp)
+	b, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return eris.Wrapf(err, "error reading sarif file %s", fileName)
+	}
+	var sarifFile bytes.Buffer
+	w := gzip.NewWriter(&sarifFile)
+	_, err = w.Write(b)
+	if err != nil {
+		return eris.Wrap(err, "error writing gzip file")
+	}
+	w.Close()
+	if len(shaResp.Sha) != 40 {
```

Comment:
A couple times when I was getting rate limited, the api sent a 200 OK response but left the SHA empty. I'm not sure why, and I haven't hit this recently, but I want to keep this check here to prevent a really annoying debugging session in the future if the error pops up more.

Reply:
ahh makes sense. perhaps we could ensure that a 200 is returned by the request, but good for now!